plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataUpdater.java

```diff
@@ -0,0 +1,252 @@
+import org.sonatype.nexus.repository.maven.internal.maven2.metadata.Maven2Metadata.Plugin;
+import org.sonatype.nexus.repository.maven.internal.maven2.metadata.Maven2Metadata.Snapshot;
+import org.sonatype.nexus.repository.storage.StorageTx;
+import org.sonatype.nexus.repository.util.TypeTokens;
+import org.sonatype.nexus.repository.view.Content;
+import org.sonatype.nexus.repository.view.payloads.BytesPayload;
+import org.sonatype.nexus.repository.view.payloads.StringPayload;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.hash.HashCode;
+import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.apache.maven.artifact.repository.metadata.SnapshotVersion;
+import org.apache.maven.artifact.repository.metadata.Versioning;
+import org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Reader;
+import org.apache.maven.artifact.repository.metadata.io.xpp3.MetadataXpp3Writer;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Maven 2 repository metadata updater.
+ *
+ * @since 3.0
+ */
+public class MetadataUpdater
+    extends ComponentSupport
+{
+  private final boolean update;
+
+  private final Repository repository;
+
+  private final MavenFacet mavenFacet;
+
+  private final Maven2MetadataMerger metadataMerger;
+
+  private final MetadataXpp3Reader metadataReader;
+
+  private final MetadataXpp3Writer metadataWriter;
+
+  public MetadataUpdater(final boolean update, final Repository repository) {
+    this.update = update;
+    this.repository = checkNotNull(repository);
+    this.mavenFacet = repository.facet(MavenFacet.class);
+    this.metadataMerger = new Maven2MetadataMerger();
+    this.metadataReader = new MetadataXpp3Reader();
+    this.metadataWriter = new MetadataXpp3Writer();
+  }
+
+  /**
+   * Processes metadata, depending on {@link #update} value and input value of metadata parameter. If input is
+   * non-null, will update or replace depending on value of {@link #update}. If update is null, will delete if {@link
+   * #update} is {@code false}.
+   */
+  public void processMetadata(final StorageTx tx, final MavenPath metadataPath, final Maven2Metadata metadata) {
+    if (metadata != null) {
+      if (update) {
+        update(tx, metadataPath, metadata);
+      }
+      else {
+        replace(tx, metadataPath, metadata);
+      }
+    }
+    else if (!update) {
+      delete(tx, metadataPath);
+    }
+  }
+
+  /**
+   * Writes/updates metadata, merges existing one, if any.
+   */
+  public void update(final StorageTx tx, final MavenPath mavenPath, final Maven2Metadata metadata) {
```

Comment:
reduce visibility of this method? the checkNotNull's could then be moved into the public processMetadata method